server/src/main/java/org/elasticsearch/rest/RestController.java

```diff
@@ -361,11 +361,12 @@ void tryAllHandlers(final RestRequest request, final RestChannel channel, final
         // Between retrieving the correct path, we need to reset the parameters,
         // otherwise parameters are parsed out of the URI that aren't actually handled.
         final Map<String, String> originalParams = new HashMap<>(request.params());
+        final Map<String, String> requestParamsRef = request.params();
         return handlers.retrieveAll(getPath(request), () -> {
             // PathTrie modifies the request, so reset the params between each iteration
-            request.params().clear();
-            request.params().putAll(originalParams);
-            return request.params();
+            requestParamsRef.clear();
```

Comment:
The way we're mutating the request itself here is still pretty crazy but we have an issue for that in https://github.com/elastic/elasticsearch/issues/25563 already